corehq/apps/linked_domain/updates.py

```diff
@@ -543,3 +553,15 @@ def _convert_reports_permissions(domain_link, master_results):
                 new_report_perms.append(get_ucr_class_name(linked_id))
 
         role_def['permissions']['view_report_list'] = new_report_perms
+
+
+def _get_new_tableau_report_permissions(downstream_domain_visualizations, upstream_permissions,
+                                        original_downstream_permissions):
+    new_downstream_view_tableau_list = []
+    for viz in downstream_domain_visualizations:
+        # Enable permissions for visualization if the viz is enabled in the linked upstream viz
+        # OR if the viz is enabled in the downstream project and doesn't have an upstream linked viz.
+        if ((viz.upstream_id and viz.upstream_id in upstream_permissions.view_tableau_list)
+        or (not viz.upstream_id and str(viz.id) in original_downstream_permissions.view_tableau_list)):
```

Comment:
nit: I sometimes find it more readable to separate these larger conditionals into their own variables. The new variable names offer an opportunity to explain what that condition represents, and potentially remove the need for comments. Don't feel like you have to change this or feel free to change this as you like, but just as an example:
```suggestion
        is_upstream_visualization = viz.upstream_id and viz.upstream_id in upstream_permissions.view_tableau_list
        is_created_by_downstream = not viz.upstream_id and str(viz.id) in original_downstream_permissions.view_tableau_list
        if is_upstream_visualization or is_created_by_downstream:
```